InfrastructureDefinition and Composite reconcilers

muvaf opened the pull request:
<!--
Thank you for helping to improve Crossplane!

We strongly recommend you look through our contributor guide at https://git.io/fj2m9
if this is your first time opening a Crossplane pull request. You can find us in
https://slack.crossplane.io/messages/dev if you need any help contributing.
-->

### Description of your changes

This PR implements the spec outlined in https://github.com/crossplane/crossplane/pull/1163 .

I am planning to get this merged and submit a separate PR for unit tests.

Fixes https://github.com/crossplane/crossplane/issues/1416

<!--
Briefly describe what this pull request does. Be sure to direct your reviewers'
attention to anything that needs special consideration.

We love pull requests that resolve an open Crossplane issue. If yours does, you
can uncomment the below line to indicate which issue your PR fixes, for example
"Fixes #500":

Fixes #
-->

### How has this code been tested?

I've manually tested the controller logic for create/update/delete mechanisms. I will test a real world scenario with Wordpress and update here.

Create the following `InfrastructureDefinition` and `Composition`:
```yaml
apiVersion: apiextensions.crossplane.io/v1alpha1
kind: InfrastructureDefinition
metadata:
  # InfrastructureDefinition names are subject to the constraints of Kubernetes
  # CustomResourceDefinition names. They must be of the form <plural>.<group>.
  name: mysqlinstances.database.example.org
spec:
  # Any composition that intends to satisfy an infrastructure resource must
  # expose each of the named connection details exactly once in any of its
  # connectionDetails objects. The connection secret published by the defined
  # infrastructure resource will include only these connection details.
  connectionSecretKeys:
    - username
    - password
    - endpoint
  # Defines the structural schema and GroupVersionKind of this infrastructure.
  # Only a single API version of the application may exist. Additional fields
  # will be injected to support composition machinery.
  crdSpecTemplate:
    group: database.example.org
    version: v1alpha1
    names:
      kind: MySQLInstance
      listKind: MySQLInstanceList
      plural: mysqlinstances
      singular: mysqlinstance
    validation:
      openAPIV3Schema:
        type: object
        properties:
          spec:
            type: object
            properties:
              engineVersion:
                type: string
              region:
                type: string
              storageGB:
                type: integer
---
apiVersion: apiextensions.crossplane.io/v1alpha1
kind: Composition
metadata:
  name: private-mysql-server
  labels:
    connectivity: private
spec:
  from:
    apiVersion: database.example.org/v1alpha1
    kind: MySQLInstance
  to:
    - base:
        apiVersion: azure.crossplane.io/v1alpha3
        kind: ResourceGroup
        spec:
          location: West US
          providerRef:
            name: example
          reclaimPolicy: Delete
      patches:
        - fromFieldPath: "spec.region"
          toFieldPath: "spec.forProvider.location"
    - base:
        apiVersion: database.azure.crossplane.io/v1beta1
        kind: MySQLServer
        spec:
          forProvider:
            administratorLogin: myadmin
            resourceGroupNameSelector:
              matchComposite: true
            location: West US
            sslEnforcement: Disabled
            version: "5.6"
            sku:
              tier: Basic
              capacity: 1
              family: Gen5
            storageProfile:
              storageMB: 20480
          writeConnectionSecretToRef:
            namespace: crossplane-system
          providerRef:
            name: example
          reclaimPolicy: Delete
      patches:
        - fromFieldPath: "metadata.uid"
          toFieldPath: "spec.writeConnectionSecretToRef.name"
        - fromFieldPath: "spec.engineVersion"
          toFieldPath: "spec.forProvider.version"
        - fromFieldPath: "spec.storageGB"
          toFieldPath: "spec.forProvider.storageMB"
          transforms:
            - type: math
              math:
                multiply: 1024
        - fromFieldPath: "spec.region"
          toFieldPath: "spec.forProvider.location"
      # Specifies the (potentially sensitive) connection details that this 'to'
      # resource should expose to the 'from' resource. Names are unique across all
      # 'to' resources within this composition. Ignored by application resources.
      connectionDetails:
        - name: username
          fromConnectionSecretKey: username
        - name: password
          fromConnectionSecretKey: password
        - name: endpoint
          fromConnectionSecretKey: endpoint
```
Then create the following CR instance:
```yaml
apiVersion: database.example.org/v1alpha1
kind: MySQLInstance
metadata:
  name: example
spec:
  engineVersion: "5.6"
  region: East US
  storageGB: 20
  compositionSelector:
    matchLabels:
      connectivity: private
```

You'll see both resources are created and reconciled.
<!--
Before reviewers can be confident in the correctness of a pull request,
it needs to tested and shown to be correct. In this section, briefly
describe the testing that has already been done or which is planned.
-->

### Checklist
<!--
Please run through the below readiness checklist. The first two items are
relevant to every Crossplane pull request.
-->
I have:
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Ensured this PR contains a neat, self documenting set of commits.
- [x] Updated any relevant [documentation] and [examples].
- [x] Reported all new error conditions into the log or as an event, as
  appropriate.

For more about what we believe makes a pull request complete, see our
[definition of done].

[documentation]: https://github.com/crossplane/crossplane/tree/master/docs
[examples]: https://github.com/crossplane/crossplane/tree/master/cluster/examples
[definition of done]: https://github.com/crossplane/crossplane/tree/master/design/one-pager-definition-of-done.md

muvaf commented:
I have just discovered a problem with deletion. Basically stopping the controller is not enough; we also have to deregister the event handler in client-go. I'll dig in to see if we can do this.

muvaf commented:
Controller is stopped when the channel is closed using @negz [fork](https://github.com/kubernetes-sigs/controller-runtime/pull/863). However, the following lines are logged when CRD that the controller was reconciling is deleted even though controller is stopped:
```
2020-04-09T22:54:49.230+0300    INFO    controller-runtime.controller   Stopping workers        {"controller": "mysqlinstances.database.example.org"}
2020-04-09T22:54:49.230+0300    INFO    crossplane      instance controller     {"controller has been stopped": "mysqlinstances.database.example.org"}
E0409 22:55:15.874863   76708 reflector.go:307] pkg/mod/k8s.io/client-go@v0.17.3/tools/cache/reflector.go:105: Failed to watch database.example.org/v1alpha1, Kind=MySQLInstance: the server could not find the requested resource
E0409 22:55:16.878131   76708 reflector.go:153] pkg/mod/k8s.io/client-go@v0.17.3/tools/cache/reflector.go:105: Failed to list database.example.org/v1alpha1, Kind=MySQLInstance: the server could not find the requested resource
E0409 22:55:17.883455   76708 reflector.go:153] pkg/mod/k8s.io/client-go@v0.17.3/tools/cache/reflector.go:105: Failed to list database.example.org/v1alpha1, Kind=MySQLInstance: the server could not find the requested resource
E0409 22:55:18.887305   76708 reflector.go:153] pkg/mod/k8s.io/client-go@v0.17.3/tools/cache/reflector.go:105: Failed to list database.example.org/v1alpha1, Kind=MySQLInstance: the server could not find the requested resource
E0409 22:55:19.890800   76708 reflector.go:153] pkg/mod/k8s.io/client-go@v0.17.3/tools/cache/reflector.go:105: Failed to list database.example.org/v1alpha1, Kind=MySQLInstance: the server could not find the requested resource
E0409 22:55:20.896285   76708 reflector.go:153] pkg/mod/k8s.io/client-go@v0.17.3/tools/cache/reflector.go:105: Failed to list database.example.org/v1alpha1, Kind=MySQLInstance: the server could not find the requested resource
E0409 22:55:21.900091   76708 reflector.go:153] pkg/mod/k8s.io/client-go@v0.17.3/tools/cache/reflector.go:105: Failed to list database.example.org/v1alpha1, Kind=MySQLInstance: the server could not find the requested resource
```

I think the [shut down call here](https://github.com/negz/controller-runtime/blob/fd19c6a/pkg/internal/controller/controller.go#L152) doesn't clean up everything. When I look at [how metacontroller stops the controller](https://github.com/GoogleCloudPlatform/metacontroller/blob/master/controller/composite/controller.go#L194), I see that it shuts down the queue, remove the event handlers from informer and then close the informer. I think the informer part is missing in Crossplane, i.e. they never handle the case where CRD is gone. So, registration of a kind is always one-way.

On the other hand, it only logs the error with no seemingly functional problem. But it's frustrating enough..

negz commented:
@muvaf I've [confirmed](https://github.com/negz/crossplane/commit/719b1b9c1f06a6a793e68c56e16e2ffa5d932efd#diff-d38c4abde00d8e13b31573e2d93132e1R157) that we can use our own cache for the watch, which allows us to stop the cache when we stop the controller, and thus avoid these "could not find the requested resource" error logs.

```go
func runExperiment(m ctrl.Manager, w v1alpha1.Watch, l logging.Logger, stop <-chan struct{}) error {
	o := controller.Options{
		Reconciler: &experimentReconciler{client: m.GetClient(), log: l, watching: w},
	}

	// TODO(negz): Set sync period?
	ca, err := cache.New(m.GetConfig(), cache.Options{
		Scheme: m.GetScheme(),
		Mapper: m.GetRESTMapper(),
	})
	if err != nil {
		return err
	}

	go func() {
		<-m.Leading()
		if err := ca.Start(stop); err != nil {
			l.Debug("cannot run experiment cache", "error", err)
		}
	}()

	ca.WaitForCacheSync(stop)

	// TODO(negz): Give this guy the same cache?
	c, err := controller.NewUnmanaged("experiment/"+w.Kind, m, o)
	if err != nil {
		return err
	}

	u := &unstructured.Unstructured{}
	u.SetAPIVersion(w.APIVersion)
	u.SetKind(w.Kind)

	if err := c.Watch(source.NewKindWithCache(u, ca), &handler.EnqueueRequestForObject{}); err != nil {
		return err
	}

	go func() {
		<-m.Leading()
		if err := c.Start(stop); err != nil {
			l.Debug("cannot run experiment controller", "error", err)
		}
	}()

	return nil
}
```

If I understand the plumbing (so much plumbing) correctly the controller-runtime `cache.Cache` is a map of informers, each for a particular GVK. So it doesn't seem _that_ bad efficiency wise that our watch would use its own informer. It does seem a little concerning that our watch is using a different cache from our controller, though. Presumably this could result in the two caches being out of sync. We could presumably hack around this by wrapping `manager.Manager` in our own implmentation whose `GetCache` method returns our desired cache, rather than the underlying manager's.

muvaf commented:
@negz thank you for digging in, it seems to be working well. I tried the cache approach but couldn't find my way around in controller-runtime logic.

>It does seem a little concerning that our watch is using a different cache from our controller, though.

I'm not sure if I understood this. For each controller started, a new cache is being initialized. It's different than `InfrastructureDefinition`'s cache but I guess this would cause problems only if they tried to work on the same resource. For now, I don't think this is a problem since the resources are expected to be managed resources, which are not cached by main Crossplane manager. But if, in the future, composition supported resources that are in main Crossplane, it might cause problem. Though I'm not much worried about this, at least for now.

muvaf commented:
@negz FWIW, it does feel awkward to call the type `InfrastructureDefinition`. Some bad ideas: `ModuleDefinition`, `GroupDefinition`, `CrossplaneResourceDefinition`, `ServiceDefinition` (maybe not so bad, but service is overloaded?)

muvaf commented:
@negz how about:
* `Composition` -> `CompositionClass`
* `InfrastructureDefinition` -> `CompositionDefinition`
* `ApplicationDefinition` -> `NamespacedCompositionDefinition` (or maybe revisit the idea of merging definitions and having a field `scope`?)

Internally we could refer to the instance as `CompositionInstance`

negz commented:
> FWIW, it does feel awkward to call the type InfrastructureDefinition. Some bad ideas: ModuleDefinition, GroupDefinition, CrossplaneResourceDefinition, ServiceDefinition (maybe not so bad, but service is overloaded?)

@muvaf Can you elaborate on why it feels awkward? Aren't we defining a new kind (or grouping) of infrastructure? All the other suggested names feel generic to the point that they don't really have meaning.

> how about Composition -> CompositionClass

I don't think any name with `Class` in it will fly. In part because (according to @bassam) user feedback has been that the class/claim pattern is not intuitive, and in part because I think `Class` implies it will only be used for initial dynamic provisioning whereas a `Composition` applies on every reconcile throughout the composed resources lifetime.

> InfrastructureDefinition -> CompositionDefinition and ApplicationDefinition -> NamespacedCompositionDefinition

It's not a generic definition of a composition though - you couldn't use it to compose a bunch of `ClusterRoles` or other generic cluster scoped resources, could you? It's specifically scoped to Crossplane "managed resources" aka "infrastructure resources".

negz commented on pkg/controller/apiextensions/composed/engine.go:
```diff
@@ -0,0 +1,111 @@
+func (c *ControllerEngine) Start(name string, gvk schema.GroupVersionKind, reconciler reconcile.Reconciler) error {
+	stop, exists := c.m[name]
+	// todo(muvaf): when a channel is closed, does it become nil? Find a way
+	// to check on the controller to see whether it crashed and needs a restart.
```
https://tour.golang.org/concurrency/4

I don't think a closed channel is `nil`, but you can read from it to test whether it's closed, e.g.:

```go
_, open := <-c.m[name]
```

Not entirely sure how this works if the channel _isn't_ closed though - presumably we don't want to block.

muvaf replied:
Tried, it blocks forever. I just wrote that part as temp fix that I will investigate later, didn't yet test it thoroughly.

negz commented on pkg/controller/apiextensions/composed/engine.go:
```diff
@@ -0,0 +1,111 @@
+	c.m[name] = stop
+	ca, err := cache.New(c.mgr.GetConfig(), cache.Options{
+		Scheme: c.mgr.GetScheme(),
+		Mapper: c.mgr.GetRESTMapper(),
```
Does this use some default resync duration if we don't plumb one in?

muvaf replied:
Yes, which is 10 hours minimum. It seems like we default to `1h` in providers but I don't see much difference honestly.

negz commented on pkg/controller/apiextensions/definer/reconciler.go:
```diff
@@ -0,0 +1,275 @@
+		definer.Status.SetConditions(runtimev1alpha1.Available())
+	}
+
+	if meta.WasDeleted(definer) {
```
I'm guessing we can't just use Kubernetes GC to remove our CRD because we need to ensure we stop the controller first?

muvaf replied:
Yes, we need to do a graceful ordered deletion like: delete instances, stop controller, delete CRD so that controller doesn't crash but it's worth a revisit later.

negz commented on pkg/controller/apiextensions/definer/api.go:
```diff
@@ -0,0 +1,109 @@
+
+// DeleteInstances deletes all instances of the generated CRD in all namespaces
+// and returns whether there are any remaining instances.
+func (m *APIDefaultCRDManager) DeleteInstances(ctx context.Context, definer CRDDefiner) (bool, error) {
```
Nit: `DeleteResources`? An "instance of a crd" is the thing defined by the resource definition, i.e. a resource.

negz replied:
Also, this is the same behaviour as deleting a CRD, right? i.e. When you delete a CRD all of the custom resources it defined are deleted?

muvaf replied:
> An "instance of a crd" is the thing defined by the resource definition, i.e. a resource.

This is correct but I think `resource` is kind of an overloaded term in the codebase and refers to `a Kubernetes resource` rather than `a Kubernetes Custom Resource`. _Deleting instances of a definer_ sounds more explicit to me.

> Also, this is the same behaviour as deleting a CRD, right? i.e. When you delete a CRD all of the custom resources it defined are deleted?

Yes, we have to delete the instances manually. It's explained in the function as:
> // NOTE(muvaf): When user deletes InfrastructureDefinition object the deletion
	// signal does not cascade to the owned resource until owner is gone. But
	// owner has its own finalizer that depends on having no instance of the CRD
	// because it cannot go away before stopping the controller.
	// So, we need to delete all instances of CRD manually here.

muvaf commented on apis/apiextensions/v1alpha1/composition_types.go:
```diff
@@ -0,0 +1,241 @@
+// should be processed.
+type TargetResourceTemplate struct {
+	// Base is the target resource that the patches will be applied on.
+	Base unstructured.Unstructured `json:"base"`
```
We only use it to copy into the target resource, which can be done via `runtime.RawExtension`

muvaf commented on apis/apiextensions/v1alpha1/instance/infrainstance_types.go:
```diff
@@ -0,0 +1,187 @@
+	_ = setObject(in, "spec.compositionRef", ref)
+}
+
+func (in *InfraInstance) GetResourceReferences() []corev1.ObjectReference {
```
Yes, I was keeping only the custom fields as `map[string]interface{}` and the rest was strong-typed. However, then I found myself doing a lot of conversions to a raw form which affects the performance:
* Every `client.Client` call had to convert.
* Every field propagation had to convert to raw form before getting the field from the resource since path could be anywhere on the resource, custom fields being raw wouldn't help.

Also, one of the main reasons that I wanted strong type was that we could generate the CRD yaml instead of having it hard-coded or manually crafted, however, that would require some investment as we can't use controller-gen as is. It could be possible but seemed cumbersome.

So, I created an interface and changed the underlying implementation to unstructured to see whether it's better and if not, basically experimenting. But I'm kind of convinced that the upsides of strong-typed version are not worth the headache it brings but I'm happy to take suggestions/arguments. I'm still in the process of restructure/refactor and all code is using interface, so we can change it easily anytime if we see fit.

muvaf commented on pkg/controller/apiextensions/composed/reconciler.go:
```diff
@@ -0,0 +1,126 @@
+
+	if err := r.ResolveSelector(ctx, cr); err != nil {
+		return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(err, "cannot resolve composition selector")
+	}
```
I couldn't see much upside of having a separate controller, which brings some overhead. Do you see how it'd benefit?

muvaf replied:
Actually, I'm thinking of using your recent selector resolver. It's pretty much the same; choose resources with given GVK using label selector to get its name. What do you think?

negz replied:
> I'm thinking of using your recent selector resolver.

Hrm, it's an unexpected usage, but it could work.

> Do you see how it'd benefit?

It's a pretty small benefit. I'd say if there was any case it would be that it would work more like our other existing controllers, and that it would give us space to build more complicated selection logic (e.g. weighted compositions to model price or performance) without adding to the complexity of this controller.

I suspect I'd like the selection to be a separate controller eventually, but it's fine to worry about that in a future PR.

muvaf replied:
I would prefer to separate it when that need comes up. We might even switch to mutating [webhook](https://github.com/crossplane/crossplane/issues/1152) for that stuff in the future.

muvaf commented on apis/apiextensions/v1alpha1/composition_types.go:
```diff
@@ -0,0 +1,241 @@
+type ConnectionDetail struct {
+	// Name of the connection secret key that will be propagated to the
+	// connection secret of the composition instance.
+	Name *string `json:"name,omitempty"`
```
Yes, I've updated the comment. I assumed that in most of the cases, name won't be changed, so, if `name` is not given we'll propagate with the same name which is `fromConnectionSecretKey`.

muvaf commented on apis/apiextensions/v1alpha1/crd_types.go:
```diff
@@ -0,0 +1,182 @@
+	// major version, then minor version. An example sorted list of versions:
+	// v10, v2, v1, v11beta2, v10beta3, v3beta1, v12alpha1, v11alpha2, foo1, foo10.
+	// +optional
+	Versions []CustomResourceDefinitionVersion `json:"versions,omitempty" protobuf:"bytes,7,rep,name=versions"`
```
I think if the user has a case where they need a conversion, they should be able to specify its config like webhook endpoint etc. But I deferred playing with this stuff for now and just used `version` field to configure the instance controller. I will revisit this logic, let's keep this conversation unresolved.

muvaf replied:
I plan to think about this in the next PR.

negz commented on pkg/controller/apiextensions/composed/engine.go:
```diff
@@ -0,0 +1,111 @@
+)
+
+// NewControllerEngine returns a new ControllerEngine instance.
+func NewControllerEngine(mgr manager.Manager, log logging.Logger) *ControllerEngine {
```
I like this type that tidies up the `NewUnmanaged` logic a little. We'll need it in the infra publication controller too; I might take a pass at breaking this out and adding it to crossplane-runtime.

muvaf commented on apis/apiextensions/v1alpha1/crds.go:
```diff
@@ -0,0 +1,287 @@
+  - Unbound
+  - Bound
+  - Released
+  type: string
```
Yeah, I'll remove binding phase. But it's worth to talk about its implications. There are two things binding phase does:
* The relation representation.
  * Per the design spec, this is handled via other means and I just left a [comment](https://github.com/crossplane/crossplane/pull/1163#discussion_r410883552) on that section.
* The readiness signal for propagation.

I think the first one is solved via other means as you pointed out but the second one is not and IMO it's necessary because the propagation starts only if the underlying resource is bound, i.e. ready, to the namespaced claim. This is a good feature since pods mounting the secret will not be started until the secret is created. However, binding phase is an overkill for this. Somehow, the namespaced publication instance should know that the cluster-scoped instance is ready. Since the CRD is defined by us, `Ready` condition would serve this purpose. @negz does that make sense to you? If so, it's worth making an addition to one-pager.

muvaf commented on apis/apiextensions/v1alpha1/composition_types.go:
```diff
@@ -0,0 +1,241 @@
+limitations under the License.
+*/
+
+package v1alpha1
```
I've updated them all.

hasheddan commented on cmd/crossplane/core/core.go:
```diff
@@ -56,6 +59,10 @@ func (c *Command) Run(log logging.Logger) error {
 	}
 
+	if err := crds.AddToScheme(mgr.GetScheme()); err != nil {
+		return errors.Wrap(err, "Cannot add Kubernetes APIs to scheme")
```
Nitpick: this made me think we were adding the core Kubernetes API types here, which I believe is done automatically as part of the `controller-runtime` machinery. We are just adding CRD and CRDList kind here right?

muvaf replied:
Kubebuilder generated code [adds the core types](https://github.com/muvaf/kubebuilder-init/blob/master/main.go#L39) via client-go scheme. Here, only CRD and CRDList kinds are added from their actual package.

hasheddan replied:
Yep, could we adjust this error message to reflect that?

muvaf replied:
Oh I didn't realize that. Fixed. Thanks for catching!

negz commented on pkg/controller/apiextensions/definer/reconciler.go:
```diff
@@ -0,0 +1,275 @@
+	}
+	// TODO(muvaf): we don't have a use case but it'd be consistent if we allowed
+	// different controller engines to be configured via ReconcilerOption.
+	r.ctrl = composite.NewControllerEngine(mgr, r.log)
```
Like a lot of our options my guess is that it would help us inject testing dependencies, even if in the real world we'd pretty much always use `composite.NewControllerEngine`.

negz commented on pkg/controller/apiextensions/definer/reconciler.go:
```diff
@@ -0,0 +1,275 @@
+		log.Debug(errGetCRD, "error", err)
+		return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(err, errGetCRD)
+	}
+	exists := !kerrors.IsNotFound(err)
```
`exists := err != nil` would also work here if we make it this far, right? Not sure whether that would be any more readable, but my brain struggled to parse `!kerrors.IsNotFound` a little due to the double negative (e.g. "is not is not found" rather than "is found").

muvaf replied:
`exists := err != nil` would be non-obvious for readers though even though we skip the `NotFound` above.

negz commented on pkg/controller/apiextensions/definer/reconciler.go:
```diff
@@ -0,0 +1,275 @@
+			definer.Status.SetConditions(runtimev1alpha1.ReconcileError(errors.Wrap(err, "cannot stop controller")))
+			return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(r.client.Status().Update(ctx, definer), errUpdateInfraDefStatus)
+		}
+		err = r.crd.Delete(ctx, definer)
+		if resource.IgnoreNotFound(err) != nil {
+			log.Debug(errDeleteCRD, "error", err)
+			definer.Status.SetConditions(runtimev1alpha1.ReconcileError(errors.Wrap(err, errDeleteCRD)))
+			return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(r.client.Status().Update(ctx, definer), errUpdateInfraDefStatus)
+		}
```
```suggestion
		if err := resource.IgnoreNotFound(r.crd.Delete(ctx, definer)); err != nil {
			log.Debug(errDeleteCRD, "error", err)
			definer.Status.SetConditions(runtimev1alpha1.ReconcileError(errors.Wrap(err, errDeleteCRD)))
			return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(r.client.Status().Update(ctx, definer), errUpdateInfraDefStatus)
		}
```

negz commented on apis/apiextensions/v1alpha1/composition_types.go:
```diff
@@ -0,0 +1,265 @@
+			return err
+		}
+		fromMap = obj
+	}
```
https://github.com/crossplane/crossplane-runtime/pull/158

muvaf replied:
Oh nice!

muvaf commented on apis/apiextensions/v1alpha1/crds.go:
```diff
@@ -0,0 +1,277 @@
+// reused, although it is known that only two different kinds will be generated.
+
+// BaseCRD returns a base template for generating a CRD.
+func BaseCRD(opts ...func(*v1beta1.CustomResourceDefinition) error) (*v1beta1.CustomResourceDefinition, error) {
```
I did that originally but since it's in `apis` folder, controller-gen returns error saying the type is not compatible. I tried to find a way to skip that type but as far as I see, you can only skip [the whole package](https://book.kubebuilder.io/reference/markers/crd.html).

muvaf commented on apis/apiextensions/v1alpha1/crds.go:
```diff
@@ -0,0 +1,277 @@
+		crd.Spec.Scope = v1beta1.ClusterScoped
+		spec := &map[string]v1beta1.JSONSchemaProps{}
+		if err := yaml.Unmarshal([]byte(InfraCompositeSpecProps), spec); err != nil {
+			return errors.Wrap(err, "constant string could not be parsed")
```
That was the case earlier. Then I thought if it's broken we won't know until later if no infradefs are shipped with Crossplane but that seems to be our future plan. So, I'll make it panic back again.

muvaf commented on apis/apiextensions/v1alpha1/crds.go:
```diff
@@ -0,0 +1,277 @@
+func InfraValidation() func(*v1beta1.CustomResourceDefinition) error {
+	return func(crd *v1beta1.CustomResourceDefinition) error {
+		crd.Spec.Scope = v1beta1.ClusterScoped
+		spec := &map[string]v1beta1.JSONSchemaProps{}
```
I plan to make this generation more modular and this will be part of that effort.

muvaf commented on apis/apiextensions/v1alpha1/infrastructuredefinition_types.go:
```diff
@@ -0,0 +1,120 @@
+// controls.
+func (in InfrastructureDefinition) GetCRDName() string {
+	return strings.ToLower(fmt.Sprintf("%s.%s", in.Spec.CRDSpecTemplate.Names.Plural, in.Spec.CRDSpecTemplate.Group))
+}
```
Yeah, that's fair.

muvaf commented on pkg/controller/apiextensions/composite/api.go:
```diff
@@ -0,0 +1,123 @@
+func NewAPIFilteredSecretPublisher(c client.Client, filter []string) *APIFilteredSecretPublisher {
+	// NOTE(negz): We transparently inject an APIPatchingApplicator in order to maintain
+	// backward compatibility with the original API of this function.
+	return &APIFilteredSecretPublisher{client: resource.NewAPIPatchingApplicator(c), filter: filter}
```
Actually, what we need is precisely `APIPatchingApplicatior` rather than any `resource.Applicatior` because the logic in the function relies on patching for partial key submission, i.e. `UpdateApplicator` would break its logic.

negz replied:
I'm sure that's true, but I still think it's better to accept an interface we can stub out for testing, rather than testing through to the implementation details of the patching applicator.

muvaf commented on pkg/controller/apiextensions/composite/composed.go:
```diff
@@ -0,0 +1,155 @@
+
+// Reconcile tries to bring the composed resource into the desired state. It
+// creates the resource if the given reference is empty.
+func (r *APIComposedReconciler) Reconcile(ctx context.Context, cr resource.Composite, composedRef v1.ObjectReference, tmpl v1alpha1.ComposedTemplate) (Observation, error) {
```
Its type is `Composition` right now but we plan to have different kinds of `Composition`s. At that point, it would be simply copy of `v1.ObjectReference`, which I'd not prefer to do.

negz replied:
I actually think we should make copies of that type, for the reasons outlined in https://github.com/crossplane/crossplane-runtime/issues/49.

muvaf commented on pkg/controller/apiextensions/composite/composed.go:
```diff
@@ -0,0 +1,155 @@
+	}
+}
+
+// APIComposedReconciler is able to reconcile a composed resource.
```
`Observe`? I don't know `Reconcile` is what it does. Happy to take suggestions there.

negz replied:
`APIComposer` with method `Compose`?

muvaf commented on pkg/controller/apiextensions/composite/composed.go:
```diff
@@ -0,0 +1,155 @@
+		if err := r.client.Get(ctx, types.NamespacedName{Name: composed.GetName(), Namespace: composed.GetNamespace()}, composed); err != nil {
+			return Observation{}, err
+		}
+	}
```
In `managed.Reconciler`, the existence of the object itself is the source of truth. But here, it's using the array of references for that. I'll update it to recreate if the resource disappeared for whatever reason but I'd like to defer that to another PR.

muvaf commented on pkg/controller/apiextensions/composite/composed.go:
```diff
@@ -0,0 +1,155 @@
+	}
+
+	// Patches are continuously applied from the Composite resource to the composed.
+	if err := r.Overlay(cr, composed, tmpl.Patches); err != nil {
```
I want to do that when I actually implement the tests.

muvaf commented on pkg/controller/apiextensions/composite/composed.go:
```diff
@@ -0,0 +1,155 @@
+
+// GetConnectionDetails returns the ConnectionDetails of the resource if a reference
+// to the connection secret exists.
+func (r *APIComposedReconciler) GetConnectionDetails(ctx context.Context, composed resource.ConnectionSecretWriterTo, filter []v1alpha1.ConnectionDetail) (managed.ConnectionDetails, error) {
```
It only returns the data rather than `Apply`ing it to the api-server. Later the output of this is filtered once more before publishing. Also `filter` sounds like a bad argument name since it contains information about conversion, so I changed it to `conversion`.

muvaf commented on pkg/controller/apiextensions/composite/reconciler.go:
```diff
@@ -0,0 +1,166 @@
+}
+
+// NewCompositeReconciler returns a new *compositeReconciler.
+func NewCompositeReconciler(name string, mgr manager.Manager, gvk schema.GroupVersionKind, log logging.Logger, filterer ConnectionSecretFilterer) reconcile.Reconciler {
```
I'll do that in the test PR.

muvaf commented on pkg/controller/apiextensions/composite/reconciler.go:
```diff
@@ -0,0 +1,166 @@
+		return reconcile.Result{}, errors.Wrap(resource.IgnoreNotFound(err), "cannot get composite resource")
+	}
+
+	if meta.WasDeleted(cr) && len(cr.GetResourceReferences()) == 0 {
```
Yes that's the beginning of the deletion process. Composed reconciler handles it.

muvaf commented on pkg/controller/apiextensions/composite/reconciler.go:
```diff
@@ -0,0 +1,166 @@
+	copy(refs, cr.GetResourceReferences())
+	conn := managed.ConnectionDetails{}
+	for i, composedRef := range refs {
+		obs, err := r.composed.Reconcile(ctx, cr, composedRef, comp.Spec.To[i])
```
Yeah I had a todo note here in the beginning about this but it would cause `resource changed` errors. In this new structure, it only returns the reference, so it should be doable. Left a note for the next PR.

muvaf commented on pkg/controller/apiextensions/composite/reconciler.go:
```diff
@@ -0,0 +1,166 @@
+
+	// We start with empty ObjectRefs and fill them up as they are provisioned.
+	refs := make([]v1.ObjectReference, len(comp.Spec.To))
+	copy(refs, cr.GetResourceReferences())
```
Yeah, that's the assumption. But honestly, I'm not too satisfied with this iteration logic. I'll try to make it better in the next PR.

muvaf commented on pkg/controller/apiextensions/definer/api.go:
```diff
@@ -0,0 +1,109 @@
+)
+
+// NewAPIInfrastructureClient returns a new APIInfrastructureClient.
+func NewAPIInfrastructureClient(client client.Client) Client {
```
It requires the `Apply` to be of type patch since there is a generation from the ground up instead of in-place changes. It creates its own `resource.ClientApplicator` to ensure that it's a patch applicator.

muvaf commented on pkg/controller/apiextensions/definer/api.go:
```diff
@@ -0,0 +1,109 @@
+	}}
+}
+
+// APIInfrastructureClient manages the generated CRD.
```
`ApplicationDefinition` might need another client for some operations like `DeleteInstances` since it deals with namespaced objects. But I'd like to leave it to `ApplicationDefinition` implementer for now.

muvaf commented on pkg/controller/apiextensions/definer/api.go:
```diff
@@ -0,0 +1,109 @@
+
+// DeleteInstances deletes all instances of the generated CRD in all namespaces
+// and returns whether there are any remaining instances.
+func (m *APIInfrastructureClient) DeleteInstances(ctx context.Context, definer Definer) (bool, error) {
```
`DeleteCustomResources` sounds good.

muvaf commented on pkg/controller/apiextensions/definer/interfaces.go:
```diff
@@ -0,0 +1,44 @@
+)
+
+// Definer has properties that are used to create a CustomResourceDefinition.
+type Definer interface {
```
I've changed the package name to be `definition`. Now it's `definition.Definer` and `definition.Client`, which I think is better.

muvaf commented:
@negz I've either left a response to your comments or resolved them by making the changes. I've added things about testing and other stuff to my list of TODOs for the next PR, which looks like this:
```
- Controller runtime tests: pkg/resource/unstructured
- External-name overrides are missing
- Forced composition selection
- Default composition
- InfraCompositeStatusProps as a map[string]v1beta1.JSONSchemaProps
- Recreate the composed resource if ref exists but resource doesn’t.
- APIComposedReconciler functions should be options so that testing is easier.
- NewReconciler functions will take params as options.
- Unit tests.
```

I've rebased this to master but it doesn't build because https://github.com/crossplane/crossplane/pull/1436 is required. Once that PR is merged, I'll rebase this and it'll be ready to merge.

muvaf commented on pkg/controller/apiextensions/composite/api.go:
```diff
@@ -0,0 +1,123 @@
+	sel := cr.GetCompositionSelector()
+	if sel == nil {
+		return errors.New("no composition selector to resolve")
+	}
```
Now that we use conditions for error propagation, it will.

muvaf commented on apis/oam/v1alpha2/core_types.go:
```diff
@@ -167,7 +167,7 @@ type ComponentSpec struct {
 	Workload runtime.RawExtension `json:"workload"`
 
-	// Parameters exposed by this component. ApplicationConfigurations that
+	// CustomFields exposed by this component. ApplicationConfigurations that
```
Oh sorry, I mistakenly thought this was the `InfraInstance` field, which doesn't exist anymore. Apparently Goland refactor touched here.

negz commented on apis/apiextensions/v1alpha1/composition_types.go:
```diff
@@ -0,0 +1,267 @@
+
+	if u, ok := to.(interface{ UnstructuredContent() map[string]interface{} }); ok {
+		return fieldpath.Pave(u.UnstructuredContent()).SetValue(c.ToFieldPath, out)
+	}
```
@muvaf I think this is another instance where the `UnstructuredContent()` call is handled by the below `ToUnstructured` call.

muvaf replied:
If it is unstructured we won't need to call `FromUnstructured` since the change is in-place. However, if it's not, we have to call `FromUnstructured` and that call does the conversion no matter what. So, it's more about whether `SetValue` operates on the actual object or the intermediary unstructured that will have to be converted back to the original struct.

muvaf commented:
@negz This should be ready to merge. I've tested it together with https://github.com/crossplane/stack-azure-sample/pull/21 and https://github.com/crossplane/app-wordpress/pull/50 and saw the light.
